File: 2016/04.cs
using System.Text.RegularExpressions;

namespace _2016;

public abstract partial class _04
{
    private static readonly string[] Data;

    static _04() => Data = Task.Run(() => Utils.GetInputData(4)).Result
        .Split('\n', StringSplitOptions.RemoveEmptyEntries);

    private static string DecryptName(string encryptedName, int sectorId) =>
        string.Concat(encryptedName.Select(ch => ch == '-' ? ' ' : (char)('a' + (ch - 'a' + sectorId) % 26)));

    private static int SolvePartOne()
    {
        var sectorIdSum = 0;

        foreach (var room in Data)
        {
            var match = SplitDetails().Match(room);
            if (!match.Success) continue;

            var encryptedName = match.Groups[1].Value;
            var sectorId = int.Parse(match.Groups[2].Value);
            var givenChecksum = match.Groups[3].Value;
            var letterCounts = new Dictionary<char, int>();

            foreach (var ch in encryptedName.Replace("-", ""))
            {
                letterCounts.TryAdd(ch, 0);
                letterCounts[ch]++;
            }

            var checksum = string.Concat(letterCounts
                .OrderByDescending(pair => pair.Value)
                .ThenBy(pair => pair.Key)
                .Take(5)
                .Select(pair => pair.Key));

            if (checksum == givenChecksum) sectorIdSum += sectorId;
        }

        return sectorIdSum;
    }

    private static int SolvePartTwo() =>
        (from room in Data
            select SplitDetails().Match(room)
            into match
            where match.Success
            let encryptedName = match.Groups[1].Value
            let sectorId = int.Parse(match.Groups[2].Value)
            let decryptedName = DecryptName(encryptedName, sectorId)
            where decryptedName.Contains("northpole object")
            select sectorId).FirstOrDefault();

    public static void Run()
    {
        Console.WriteLine($"Part One: {SolvePartOne()}");
        Console.WriteLine($"Part Two: {SolvePartTwo()}");
    }

    [GeneratedRegex(@"([a-z-]+)(\d+)\[([a-z]+)\]")]
    private static partial Regex SplitDetails();
}

File: 2017/04.fs
namespace _2017

open System

module _04 =
    let Data = (Utils.GetInputData 4).Split('\n', StringSplitOptions.RemoveEmptyEntries)

    let isValid(anagramFree: bool) (passphrase: string) =
        let words = passphrase.Split(' ')
        let normalizedWords =
            if anagramFree then
                words |> Array.map (fun word -> word.ToCharArray() |> Array.sort |> String)
            else
                words
        let uniqueWords = normalizedWords |> Set.ofArray
        Array.length normalizedWords = Set.count uniqueWords

    let solvePartOne() = Data |> Array.filter (isValid false) |> Array.length

    let solvePartTwo() = Data |> Array.filter (isValid true) |> Array.length

    let Run() =
        printfn $"Part One: {solvePartOne()}"
        printfn $"Part Two: {solvePartTwo()}"